New Lint: Pass small trivially copyable objects by value

Nemo157 opened the pull request:
Fixes #1680

Hardcoded for 64-bit "trivial" size for now, suggestions on whether this should be target-specific/configurable/something else would be appreciated.

Nemo157 commented:
Discussed configuration with @oli-obk and made it configurable in the config file, or default to 2 * target register size limited to 8 bytes max.

Not certain that the configuration macro supports `Option<u64>`, but I can't see any existing tests for changing configuration options to base tests for that on.

flip1995 commented on README.md:
```diff
@@ -7,7 +7,7 @@
 A collection of lints to catch common mistakes and improve your [Rust](https://github.com/rust-lang/rust) code.
 
-[There are 259 lints included in this crate!](https://rust-lang-nursery.github.io/rust-clippy/master/index.html)
+[There are 260 lints included in this crate!](https://rust-lang-nursery.github.io/rust-clippy/master/index.html)
```
I guess this (and some other stuff) got changed by the `update_lints.py` script. That's ok for now, but it could lead to some annoying rebasing on your side, if someone else changes this too. Consider to only add things in your PR, which is important for your lint/fix/... and make an additional PR for the `update_lints.py` script. It may also save you some work :wink:

flip1995 commented on clippy_lints/src/trivially_copy_pass_by_ref.rs:
```diff
@@ -0,0 +1,135 @@
+        cx: &LateContext<'a, 'tcx>,
+        kind: FnKind<'tcx>,
+        decl: &'tcx FnDecl,
+        _body: &'tcx Body,
```
Imo `_` looks nicer than `_body` here.

flip1995 commented on clippy_lints/src/trivially_copy_pass_by_ref.rs:
```diff
@@ -0,0 +1,135 @@
+                        TRIVIALLY_COPY_PASS_BY_REF,
+                        input.span,
+                        "this argument is passed by reference, but would be more efficient if passed by value",
+                        "consider taking by value instead",
```
Is `"consider passing by value instead"` maybe better here?

flip1995 commented:
> Not certain that the configuration macro supports Option\<u64\>, but I can't see any existing tests for changing configuration options to base tests for that on.

I think you can make a new directory like `tests/ui/toml_trivially_copy` and create a `clippy.toml` with `trivial-copy-size-limit = Some(16)` there. Then you can add tests in this directory with the given configuration

killercup commented on clippy_lints/src/trivially_copy_pass_by_ref.rs:
```diff
@@ -0,0 +1,135 @@
+                // it to 2 * the register size, capped at 32-bit to retain
+                // portability between 32 and 64-bit.
+                let bit_width = cx.sess().target.usize_ty.bit_width().unwrap_or(32);
+                cmp::min(bit_width as u64 / 8, 4) * 2
```
I don't think the doc comment above is right. aren't you capping it at 64 bits? `min(x, 4) * 2` is ≤ 8 bytes = 64 bit

Nemo157 replied:
I'll try and reword the comment a bit, it caps the _register size_ at 32-bits, then doubles that to get the maximum size of a type that should be passed by value.

Nemo157 commented:
Addressed the individual comments. I've added UI tests for configuration that depend on #2825, will probably need a small update once that's sorted out. I'll try and do the dogfood refactor tonight as well, but might have to wait till tomorrow.

Nemo157 commented:
Hmm, taking a glance at the errors shows [an issue with the suggestion](https://travis-ci.org/rust-lang-nursery/rust-clippy/jobs/384377244#L987-L993)

```
error: this argument is passed by reference, but would be more efficient if passed by value
   --> clippy_lints/src/attrs.rs:263:52
    |
263 | fn check_attrs(cx: &LateContext, span: Span, name: &Name, attrs: &[Attribute]) {
    |                                                    ^^^^^ help: consider taking by value instead: `syntax_pos::symbol::Symbol`
    |
    = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.205/index.html#trivially_copy_pass_by_ref
```

I guess it should probably be using the type from the `FnDecl::inputs` instead of the `FnSig::inputs` to keep it consistent with what was written.

flip1995 commented on clippy_lints/src/utils/conf.rs:
```diff
@@ -156,6 +156,8 @@ define_Conf! {
     /// Lint: DECIMAL_LITERAL_REPRESENTATION. The lower bound for linting decimal literals
     (literal_representation_threshold, "literal_representation_threshold", 16384 => u64),
+    /// Lint: TRIVIALLY_COPY_PASS_BY_REF. The maximum size to consider a copy type as "trivial".
+    (trivial_copy_size_limit, "trivial_copy_size_limit", None => Option<u64>),
```
Oh this configuration is in bytes not in bits. I thought it was in bits for no good reason 😃 . Could you add `(in bytes)` or something similar to the documentation?

flip1995 commented:
> I guess it should probably be using the type from the FnDecl::inputs instead of the FnSig::inputs to keep it consistent with what was written.

I guess this would also be better in the case of the `self` argument [here](https://travis-ci.org/rust-lang-nursery/rust-clippy/jobs/384377244#L1043-L1049)

flip1995 commented on clippy_lints/src/trivially_copy_pass_by_ref.rs:
```diff
@@ -0,0 +1,148 @@
+                            input.span,
+                            "this argument is passed by reference, but would be more efficient if passed by value",
+                            "consider passing by value instead",
+                            snippet(cx, decl_ty.span, "_").into());
```
The calls of the `span_lint_and_sugg` functions are exactly the same, except the `sugg` argument. Could you prettify this? Everything else lgtm.

oli-obk commented on clippy_lints/src/trivially_copy_pass_by_ref.rs:
```diff
@@ -0,0 +1,143 @@
+            }
+
+            let limit = self.limit.unwrap_or_else(|| {
+                let bit_width = cx.sess().target.usize_ty.bit_width().unwrap_or(32) as u64;
```
`bit_width` returning `None` is a bug. Just `expect` this here.

oli-obk commented on clippy_lints/src/trivially_copy_pass_by_ref.rs:
```diff
@@ -0,0 +1,143 @@
+                return;
+            }
+
+            let limit = self.limit.unwrap_or_else(|| {
```
Move this to the `new` method so it's only done once

oli-obk commented on clippy_lints/src/utils/conf.rs:
```diff
@@ -156,6 +156,8 @@ define_Conf! {
     (verbose_bit_mask_threshold, "verbose_bit_mask_threshold", 1 => u64),
     /// Lint: DECIMAL_LITERAL_REPRESENTATION. The lower bound for linting decimal literals
     (literal_representation_threshold, "literal_representation_threshold", 16384 => u64),
+    /// Lint: TRIVIALLY_COPY_PASS_BY_REF. The maximum size (in bytes) to consider a copy type as "trivial".
```
"copy type" => "`Copy` type"

"as trivial" => "for passing by value instead of by reference"

flip1995 commented:
Ping @oli-obk

oli-obk commented:
You do have merge powers @flip1995 😝

Nemo157 commented:
Rebased with the fixup commit squashed in

flip1995 commented:
Well let's use them than. :smile:
